## 七月在线机器学习集训营八期第六周(机器学习)考试
#### 考试说明:
- 起止时间：请同学在2019年06月28日至07月01日期间完成，最晚提交时间下周一（07月1日12时之前）结束，<b>逾期不接受补考,该考试分数计入平时成绩</b>
- 考试提交方式：请同学<font color=red><b>拷贝</b></font>该试卷后，将文件更名为同学姓名拼音-exam6（例如wangwei-exam6）后，移动至/0.Teacher/Exam/6/目录下后，再进行作答。
- 注意事项：为确保同学们真正了解自身对本周课程的掌握程度，<font color=red><b>请勿翻阅抄袭，移动，更改</b></font>其它同学试卷。如发现按0分处理
- 请同学在下方同学姓名处填写自己的姓名，批改人和最终得分处不用填写


- 同学姓名: 王迪
- 批改人： 
- 最终得分:

<center><h1>####答卷开始####</h1></center>
***

## 简答题(共4题，每题10分，共计40分)

#### 1. 请写出你了解的机器学习特征工程操作，以及它的意义（面试题）

特征工程是一项从数据中提取特征，然后转换成适合机器学习模型的格式的科学工程。（甚至是一种艺术）

特征工程包括三大块任务：数据采集，数据和特征处理，特征选择  

一、数据采集，需要考虑数据的可获取性，有效性等  

二、数据和特征处理

数据处理主要分为数据清洗和数据采样。  
数据清洗需要考虑缺失值和脏数据的处理。比如需要删除掉某些异常数据、噪声数据。采用何种策略填补缺省值。  
数据采样要考虑正负样本的均衡性。采样方法有随机采样、分层采样、上采样（内插值），下采样（缩放）和过采样（oversampling）  

特征处理

根据特征的类型，对特征值进行处理。  
数值型。1，对特征幅度进行处理，比如归一化、标准化方法。2. 离散化处理。将连续值切分（等距离切分、等分位数切分）  
类别型。有one-hot编码、hash技巧等  
时间型。时间可以处理为连续型（如持续时间长度，间隔时间长度等），也可处理为离散型（如周几、哪个时间段）  
文本型。可将文本数据处理为稀疏矩阵等技巧  
统计型。最大值、最小值、分位数、比例、次序等数据  
组合特征。

三、特征选择

由于存在特征共线性、无用特征、过拟合等问题，需要进行特征选择。

特征选择的方式：
1. 过滤法,根据方差大小和与结果的相关度，排序留下最重要的那些特征。
2. 包装法，每次选择或排除若干特征，选择最佳特征子集。搜索特征子集，评估模型效果，进而筛选出需要的特征子集。
3. Embedded集成法/嵌入法：通过机器学习得到各个特征的权值系数。分析特征的重要性（比如L1正则）


意义
数据和特征决定了机器学的上限，而模型和算法只是逼近这个上限而已。

正确的特征可以减少模型的复杂度，并因此使模型能输出更好的结果。




#### 2.请写出上述特征工程操作的sklearn或者pandas实现方式


数据清洗需要考虑缺失值和脏数据的处理。
这里可以用pandas 的fillna方法填充缺失值。可以基于马氏距离检测离群点。

数据采样。pandas里有sample方法

特征处理。
pandas的cut方法可以将数值型分段。get_dummy方法可以将类别型特征编成one-hot编码。
pandas对时间、文本型数据也有比较好的支持。

sklearn.preproccessing库的StandardScaler类可以对数值型做标准化处理；MinMaxScale做归一化处理。OneHotEncoder类也可以做one-hot编码


特征选择。  
Filter过滤法，sklearn.feature_selection库的VarianceThreshold类可以通过设定阈值筛选  
卡方检验筛选或者互信息熵。sklearn.feature_selection库的SelectKBest类。

Wrapper包装法/包装法  
比如sklearn.feature_selection 类的 RFE。利用递归特征消除算法寻找最佳特征子集。

Embedded集成法/嵌入法  
比如基于L1、L2正则化的特征选择法。feature_selection库的SelectFromModel类结合sklearn.linear_model库的 LogisticRegression逻辑回归类

比如基于树模型的特征选择法。



#### 3.模型评估中的留一法，留出法，交叉验证分别是什么操作？

留出法：直接将数据集$D$划分为两个互斥的集合，其中一个集合作为训练集$S$,另一个作为测试集$T$，在$S$上训练出模型后，用$T$来评估其测试误差，作为对泛化误差的估计。

交叉验证法：将数据集$D$划分为$k$个大小相似的互斥子集，$D=D_1 \cup D_2 \cup \cdots \cup D_k$， 其中$D_i$从$D$中分层采样抽样出来。
每次用$k-1$个子集的并集作为训练集，余下的子集作为测试集。这样得到$k$组训练测试集。


留一法：假设$D$中有$m$个样本，若令$k=m$，则得到了交叉验证法的一个特例：留一法。但是留一法在数据集比较大时，计算开销较大。



#### 4.如何理解模型的过拟合与欠拟合，以及如何解决？（面试题）

过拟合模型在数据集上具有高方差（high variance）和低偏差（low bias）。对应表现为过拟合模型在训练集上效果较好，但是在测试集上表现较差。
方差是模型响应训练数据而变化的程度，偏差是模型

欠拟合模型在数据集上具有高偏差。对应表现为过拟合模型在训练集上效果就很差。

解决欠拟合问题，往往需要添加新的特征，比如引入更复杂的模型（采用非线性模型、深度神经网络），另外，还可以采用boosting算法可以不断提升对残差的拟合。

解决过拟合问题，一般可以进行正则化，提前停止法，还可以采用平均值法，bagging，降维等等。

## 操作题(共1题，共计60分)

### 信用卡欺诈项目(共6项，每项10分)

 #### 前期为数据导入,预览及粗略的预处理(此部分请勿修改，涉及的数据文件无需复制移动。实验题在下方)

In [1]:
# import pandas as pd
# pd.set_option('display.max_columns', 500)
# import zipfile
# with zipfile.ZipFile('http://ml8.julyedu.com:8901/edit/0.Teacher/Data/KaggleCredit2.zip', 'r') as z:
#     f = z.open('KaggleCredit2.csv')
#     data = pd.read_csv(f, index_col=0)
# data.head()


In [2]:
import pandas as pd
pd.set_option('display.max_columns', 500)
import zipfile
# with zipfile.ZipFile('../../Data/KaggleCredit2.zip', 'r') as z:
with zipfile.ZipFile('data/KaggleCredit2.zip', 'r') as z:
    f = z.open('KaggleCredit2.csv')
    data = pd.read_csv(f, index_col=0)
data.head()

SeriousDlqin2yrs  RevolvingUtilizationOfUnsecuredLines   age  \
0                 1                              0.766127  45.0   
1                 0                              0.957151  40.0   
2                 0                              0.658180  38.0   
3                 0                              0.233810  30.0   
4                 0                              0.907239  49.0   

   NumberOfTime30-59DaysPastDueNotWorse  DebtRatio  MonthlyIncome  \
0                                   2.0   0.802982         9120.0   
1                                   0.0   0.121876         2600.0   
2                                   1.0   0.085113         3042.0   
3                                   0.0   0.036050         3300.0   
4                                   1.0   0.024926        63588.0   

   NumberOfOpenCreditLinesAndLoans  NumberOfTimes90DaysLate  \
0                             13.0                      0.0   
1                              4.0                      0.0   
2                              2.0                      1.0   
3                              5.0                      0.0   
4                              7.0                      0.0   

   NumberRealEstateLoansOrLines  NumberOfTime60-89DaysPastDueNotWorse  \
0                           6.0                                   0.0   
1                           0.0                                   0.0   
2                           0.0                                   0.0   
3                           0.0                                   0.0   
4                           1.0                                   0.0   

   NumberOfDependents  
0                 2.0  
1                 1.0  
2                 0.0  
3                 0.0  
4                 0.0

In [3]:
data.shape

(112915, 11)

In [4]:
data.isnull().sum(axis=0)

SeriousDlqin2yrs                           0
RevolvingUtilizationOfUnsecuredLines       0
age                                     4267
NumberOfTime30-59DaysPastDueNotWorse       0
DebtRatio                                  0
MonthlyIncome                              0
NumberOfOpenCreditLinesAndLoans            0
NumberOfTimes90DaysLate                    0
NumberRealEstateLoansOrLines               0
NumberOfTime60-89DaysPastDueNotWorse       0
NumberOfDependents                      4267
dtype: int64

In [5]:
data.dropna(inplace=True)
data.shapey = data['SeriousDlqin2yrs']
X = data.drop('SeriousDlqin2yrs', axis=1)

D:\anaconda3\lib\site-packages\ipykernel_launcher.py:2: UserWarning: Pandas doesn't allow columns to be created via a new attribute name - see https://pandas.pydata.org/pandas-docs/stable/indexing.html#attribute-access
  


In [6]:
y = data['SeriousDlqin2yrs']
X = data.drop('SeriousDlqin2yrs', axis=1)
y.mean()

0.06742876076872101

### 以下为操作要求，每题10分。
### 注意：每题都要附加上对实验过程和结果的总结，否则没有分数

#### 1.把数据切分成训练集和测试集后，使用logistic regression建模，并且输出一下系数，分析重要度。  

In [7]:
from sklearn.preprocessing import StandardScaler
import warnings
warnings.filterwarnings("ignore")

#### 数据整体分析

In [8]:
X.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 108648 entries, 0 to 112914
Data columns (total 10 columns):
RevolvingUtilizationOfUnsecuredLines    108648 non-null float64
age                                     108648 non-null float64
NumberOfTime30-59DaysPastDueNotWorse    108648 non-null float64
DebtRatio                               108648 non-null float64
MonthlyIncome                           108648 non-null float64
NumberOfOpenCreditLinesAndLoans         108648 non-null float64
NumberOfTimes90DaysLate                 108648 non-null float64
NumberRealEstateLoansOrLines            108648 non-null float64
NumberOfTime60-89DaysPastDueNotWorse    108648 non-null float64
NumberOfDependents                      108648 non-null float64
dtypes: float64(10)
memory usage: 9.1 MB


In [9]:
X.describe()

RevolvingUtilizationOfUnsecuredLines            age  \
count                         108648.000000  108648.000000   
mean                               5.989695      51.361304   
std                              259.792416      14.453571   
min                                0.000000       0.000000   
25%                                0.034319      41.000000   
50%                                0.172905      51.000000   
75%                                0.570881      62.000000   
max                            50708.000000     103.000000   

       NumberOfTime30-59DaysPastDueNotWorse      DebtRatio  MonthlyIncome  \
count                         108648.000000  108648.000000   1.086480e+05   
mean                               0.378636       0.306237   6.964368e+03   
std                                3.525280       0.222924   1.492419e+04   
min                                0.000000       0.000000   1.000000e+00   
25%                                0.000000       0.133493   3.640000e+03   
50%                                0.000000       0.278311   5.600000e+03   
75%                                0.000000       0.440113   8.416000e+03   
max                               98.000000       0.999909   3.008750e+06   

       NumberOfOpenCreditLinesAndLoans  NumberOfTimes90DaysLate  \
count                    108648.000000            108648.000000   
mean                          8.679055                 0.213488   
std                           5.128120                 3.493366   
min                           0.000000                 0.000000   
25%                           5.000000                 0.000000   
50%                           8.000000                 0.000000   
75%                          11.000000                 0.000000   
max                          57.000000                98.000000   

       NumberRealEstateLoansOrLines  NumberOfTime60-89DaysPastDueNotWorse  \
count                 108648.000000                         108648.000000   
mean                       1.015325                              0.188664   
std                        1.077678                              3.476195   
min                        0.000000                              0.000000   
25%                        0.000000                              0.000000   
50%                        1.000000                              0.000000   
75%                        2.000000                              0.000000   
max                       29.000000                             98.000000   

       NumberOfDependents  
count       108648.000000  
mean             0.856574  
std              1.149537  
min              0.000000  
25%              0.000000  
50%              0.000000  
75%              2.000000  
max             20.000000

#### 删除掉离群点

In [10]:
# 删除掉离群特征点
def remove_outlier(df):
    df_new = df.copy()
    for col in ['RevolvingUtilizationOfUnsecuredLines',
       'NumberOfTime30-59DaysPastDueNotWorse', 'DebtRatio',
       'NumberOfOpenCreditLinesAndLoans', 'NumberOfTimes90DaysLate',
       'NumberRealEstateLoansOrLines', 'NumberOfTime60-89DaysPastDueNotWorse',
       'NumberOfDependents']:
        rule1 = df_new[col] < df_new[col].quantile(q=0.999)
        df_new = df_new[rule1]
    return df_new

X_ = remove_outlier(X)
X_.describe()

RevolvingUtilizationOfUnsecuredLines            age  \
count                         107395.000000  107395.000000   
mean                               0.562973      51.408492   
std                               13.693150      14.468018   
min                                0.000000       0.000000   
25%                                0.033859      41.000000   
50%                                0.169532      51.000000   
75%                                0.560595      62.000000   
max                             1179.000000     103.000000   

       NumberOfTime30-59DaysPastDueNotWorse      DebtRatio  MonthlyIncome  \
count                         107395.000000  107395.000000   1.073950e+05   
mean                               0.249537       0.305123   6.947284e+03   
std                                0.690212       0.221494   1.495111e+04   
min                                0.000000       0.000000   1.000000e+00   
25%                                0.000000       0.133273   3.642000e+03   
50%                                0.000000       0.277817   5.600000e+03   
75%                                0.000000       0.438548   8.400000e+03   
max                               13.000000       0.990234   3.008750e+06   

       NumberOfOpenCreditLinesAndLoans  NumberOfTimes90DaysLate  \
count                    107395.000000            107395.000000   
mean                          8.644239                 0.079361   
std                           4.991891                 0.396488   
min                           0.000000                 0.000000   
25%                           5.000000                 0.000000   
50%                           8.000000                 0.000000   
75%                          11.000000                 0.000000   
max                          34.000000                 5.000000   

       NumberRealEstateLoansOrLines  NumberOfTime60-89DaysPastDueNotWorse  \
count                 107395.000000                         107395.000000   
mean                       1.003063                              0.054798   
std                        1.011977                              0.258251   
min                        0.000000                              0.000000   
25%                        0.000000                              0.000000   
50%                        1.000000                              0.000000   
75%                        2.000000                              0.000000   
max                        7.000000                              2.000000   

       NumberOfDependents  
count       107395.000000  
mean             0.844592  
std              1.120570  
min              0.000000  
25%              0.000000  
50%              0.000000  
75%              2.000000  
max              5.000000

In [11]:
y = y[X_.index].reset_index(drop=True)

#### 年龄、收入分箱分析

In [12]:
# 注意到age有为0的，给以平均值
X.loc[X.age == 0, "age"] = X.age.mean()

cat_age = pd.qcut(X.age, [0, .33, .67, 1.])
cat_age.cat.categories
# 对age数值型特征 进行三分箱，大约为；[20-40, 40-60, 60+]；对应青年、中年、老年用户

IntervalIndex([(20.999, 44.0], (44.0, 58.0], (58.0, 103.0]],
              closed='right',
              dtype='interval[float64]')

In [13]:
cat_income = pd.qcut(X.MonthlyIncome, [0, .25, 0.5, .75, 1.])
cat_income.cat.categories
# 对月收入进行四分箱。
# [0,3500], [3500, 5500], [5500, 8500], [8500,+] 分别对应低收入、中低收入、中高收入、高收入群体

IntervalIndex([(0.999, 3640.0], (3640.0, 5600.0], (5600.0, 8416.0], (8416.0, 3008750.0]],
              closed='right',
              dtype='interval[float64]')

#### 年龄、收入变换为分类型特征

In [14]:
def data_preprocessing(df):
    num_attr=df[['age','MonthlyIncome']]
    X_cat = pd.DataFrame(columns=['age','month_income'])
    X_cat['age'] = pd.cut(num_attr['age'],[0,40,60,120],labels=['20-40','40-60','60+'])
    X_cat['month_income'] = pd.cut(num_attr['MonthlyIncome'],[0,3500,5500,8500,10000000],labels=['0-3500','3500-5500','5500-8500','8500+'])
    cat_attr=pd.get_dummies(X_cat)
    return cat_attr

X_cat=data_preprocessing(X_)
X_cat.head()

age_20-40  age_40-60  age_60+  month_income_0-3500  month_income_3500-5500  \
0          0          1        0                    0                       0   
1          1          0        0                    1                       0   
2          1          0        0                    1                       0   
3          1          0        0                    1                       0   
4          0          1        0                    0                       0   

   month_income_5500-8500  month_income_8500+  
0                       0                   1  
1                       0                   0  
2                       0                   0  
3                       0                   0  
4                       0                   1

#### 其他变量标准化

In [15]:

def standardize(df):
    scaler = StandardScaler()
    scaler.fit(df)
    arr = scaler.transform(df)
    return pd.DataFrame(arr, columns=df.columns)



X_num = standardize(X_[['RevolvingUtilizationOfUnsecuredLines',
       'NumberOfTime30-59DaysPastDueNotWorse', 'DebtRatio',
       'NumberOfOpenCreditLinesAndLoans', 'NumberOfTimes90DaysLate',
       'NumberRealEstateLoansOrLines', 'NumberOfTime60-89DaysPastDueNotWorse',
       'NumberOfDependents']])
X_num.head()

RevolvingUtilizationOfUnsecuredLines  NumberOfTime30-59DaysPastDueNotWorse  \
0                              0.014836                              2.536137   
1                              0.028787                             -0.361538   
2                              0.006953                              1.087299   
3                             -0.024039                             -0.361538   
4                              0.025142                              1.087299   

   DebtRatio  NumberOfOpenCreditLinesAndLoans  NumberOfTimes90DaysLate  \
0   2.247741                         0.872571                -0.200162   
1  -0.827322                        -0.930361                -0.200162   
2  -0.993300                        -1.331013                 2.321997   
3  -1.214813                        -0.730035                -0.200162   
4  -1.265036                        -0.329383                -0.200162   

   NumberRealEstateLoansOrLines  NumberOfTime60-89DaysPastDueNotWorse  \
0                      4.937818                             -0.212188   
1                     -0.991196                             -0.212188   
2                     -0.991196                             -0.212188   
3                     -0.991196                             -0.212188   
4                     -0.003027                             -0.212188   

   NumberOfDependents  
0            1.031094  
1            0.138687  
2           -0.753720  
3           -0.753720  
4           -0.753720

In [16]:
print(X_num.shape)
print(X_cat.shape)

(107395, 8)
(107395, 7)


#### 组合处理后的数值型、类别型特征

In [17]:
X_new = pd.concat([X_num.reset_index(drop=True), X_cat.reset_index(drop=True)], axis=1)

#### 开始训练模型

In [18]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X_new, y, test_size=0.33, random_state=42)

In [19]:
# 考虑样本不平衡问题
from sklearn.linear_model import LogisticRegression
lr_model = LogisticRegression(class_weight='balanced')
lr_model.fit(X_train, y_train)

LogisticRegression(C=1.0, class_weight='balanced', dual=False,
          fit_intercept=True, intercept_scaling=1, max_iter=100,
          multi_class='warn', n_jobs=None, penalty='l2', random_state=None,
          solver='warn', tol=0.0001, verbose=0, warm_start=False)

In [20]:
print(lr_model.coef_)
print(lr_model.intercept_)

[[ 0.02286012  0.49193208  0.24043842  0.01324506  0.5249304  -0.09487894
   0.2921626   0.06488568  0.37062551  0.06680945 -0.52687605  0.14417785
  -0.04379487 -0.19778421 -0.25029235]]
[-0.34769357]


In [21]:
print(X_train.columns)

Index(['RevolvingUtilizationOfUnsecuredLines',
       'NumberOfTime30-59DaysPastDueNotWorse', 'DebtRatio',
       'NumberOfOpenCreditLinesAndLoans', 'NumberOfTimes90DaysLate',
       'NumberRealEstateLoansOrLines', 'NumberOfTime60-89DaysPastDueNotWorse',
       'NumberOfDependents', 'age_20-40', 'age_40-60', 'age_60+',
       'month_income_0-3500', 'month_income_3500-5500',
       'month_income_5500-8500', 'month_income_8500+'],
      dtype='object')


由上可以知道，NumberOfTime30-59DaysPastDueNotWorse，NumberOfTimes90DaysLate，age_60+，age_20-40，NumberOfTime60-89DaysPastDueNotWorse等特征比较重要

####  模型评估

In [22]:
from sklearn.model_selection import cross_val_score
cross_val_score(lr_model, X_train, y_train, cv=5, scoring='accuracy')

array([0.83198999, 0.83051907, 0.8321173 , 0.8323836 , 0.82696317])

In [23]:
cross_val_score(lr_model, X_train, y_train, cv=5, scoring='roc_auc')

array([0.79022833, 0.81692512, 0.80641439, 0.80532066, 0.79735062])

#### 2.使用决策树/SVM/KNN...等sklearn分类算法进行分类，尝试了解参数含义，调整不同的参数。

In [24]:
from sklearn.model_selection import GridSearchCV
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import LinearSVC
from sklearn.neighbors import KNeighborsClassifier

#### 决策树

In [25]:
param_grid={"max_depth":[2, 4, 8], "class_weight": ['balanced']}  
# max_depth 树的深度
dt = GridSearchCV(DecisionTreeClassifier(),param_grid=param_grid,cv=5,scoring='roc_auc')
dt.fit(X_train, y_train)
print(dt.best_params_, dt.best_score_)

{'class_weight': 'balanced', 'max_depth': 4} 0.8185185701333904


#### SVM

In [26]:
param_grid={"C":[0.1, 1, 10], "dual": [True,False],"class_weight":['balanced']} 
# C :惩罚系数 dual:是否使用对偶形式来优化算法 class_weight:样本权重
svm = GridSearchCV(LinearSVC(),param_grid=param_grid,cv=5,scoring='roc_auc')
svm.fit(X_train, y_train)
print(svm.best_params_, svm.best_score_)

#### KNN

In [27]:
param_grid={"n_neighbors":[3, 5], "weights": ['distance'],"metric":['euclidean']}
# n_neighbors:KNN算法的K值
#weights：每个样本的近邻样本的权重  "distance"，则权重和距离成反比例
#metric：距离，欧氏距离
knn = GridSearchCV(KNeighborsClassifier(),param_grid=param_grid,cv=5,scoring='roc_auc')
knn.fit(X_train, y_train)
print(svm.best_params_, svm.best_score_)

结论：
决策树的AUC最高,且训练时间上，SVM和KNN的训练时间比LR和决策树多，并且可解释性上也比LR和决策树差。  
所以在此样本集上，LR和决策树效果要比SVM、KNN好。  
因为正负样本不平衡，应当用AUC来评估模型。  

#### 3.在测试集上进行预测，计算准确度

In [28]:
from sklearn.metrics import accuracy_score, roc_auc_score
#LR
accuracy_score(y_test, lr_model.predict(X_test))
roc_auc_score(y_test, lr_model.predict(X_test))

0.7499452565707556

In [29]:
#决策树
accuracy_score(y_test, lr_model.predict(X_test))
roc_auc_score(y_test,dt.best_estimator_.predict(X_test))

0.7559380244490705

####  4.查看sklearn的官方说明，了解混淆矩阵等评估标准，并对此例进行评估。

In [30]:
from sklearn.metrics import confusion_matrix
con_matrix = confusion_matrix(y_test, lr_model.predict(X_test))
con_matrix

array([[28089,  5098],
       [  781,  1473]], dtype=int64)

 预测  非欺诈   欺诈
 
 真实非欺诈 28089  5098
 真实欺诈  781    1473

召回率 tpr = 65.3%，说明还有近35%的欺诈者没有被召回。

####  5.银行通常会有更严格的要求，因为fraud带来的后果通常比较严重，一般我们会调整模型的标准。 请尝试调整阈值，并观察，记录这些变化。
> 比如在logistic regression当中，一般我们的概率判定边界为0.5，但是我们可以把阈值设定低一些，来提高模型的“敏感度”   
试试看把阈值设定为0.3，再看看这个时候的混淆矩阵等评估指标。

In [31]:
def my_matrix(x_test,y_test,threshold):
    y_pred = lr_model.predict_proba(x_test)[:,1]>threshold
    con_matrix = confusion_matrix(y_test,y_pred)
    return con_matrix

my_matrix(X_test,y_test,threshold=0.3)


array([[13189, 19998],
       [  179,  2075]], dtype=int64)

召回率 tpr = 85.2%，相对于0.5的阈值，召回率提升了20个百分点。降低阈值可以提高模型的召回率，也即所谓的敏感度。

#### 6.尝试对不同特征的重要度进行排序，通过特征选择的方式，对特征进行筛选。并重新建模，观察此时的模型准确率等评估指标。

In [32]:
from sklearn.feature_selection import RFE
lr = LogisticRegression(class_weight='balanced')
rfe = RFE(lr,n_features_to_select=5)
rfe.fit(X_train,y_train)

RFE(estimator=LogisticRegression(C=1.0, class_weight='balanced', dual=False,
          fit_intercept=True, intercept_scaling=1, max_iter=100,
          multi_class='warn', n_jobs=None, penalty='l2', random_state=None,
          solver='warn', tol=0.0001, verbose=0, warm_start=False),
  n_features_to_select=5, step=1, verbose=0)

In [33]:
sorted(list(zip(rfe.ranking_,X_train.columns)))

[(1, 'NumberOfTime30-59DaysPastDueNotWorse'),
 (1, 'NumberOfTime60-89DaysPastDueNotWorse'),
 (1, 'NumberOfTimes90DaysLate'),
 (1, 'age_60+'),
 (1, 'month_income_0-3500'),
 (2, 'age_20-40'),
 (3, 'DebtRatio'),
 (4, 'month_income_8500+'),
 (5, 'month_income_5500-8500'),
 (6, 'NumberRealEstateLoansOrLines'),
 (7, 'NumberOfDependents'),
 (8, 'age_40-60'),
 (9, 'month_income_3500-5500'),
 (10, 'RevolvingUtilizationOfUnsecuredLines'),
 (11, 'NumberOfOpenCreditLinesAndLoans')]

较重要的特征是(1, 'NumberOfTime30-59DaysPastDueNotWorse'),
 (1, 'NumberOfTime60-89DaysPastDueNotWorse'),
 (1, 'NumberOfTimes90DaysLate'),
 (1, 'age_60+'),
 (1, 'month_income_0-3500')
与之前的观察基本吻合

***
<center><h1>####答卷结束####</h1></center>

### 本周课程意见反馈(非必答)
请同学围绕以下两点进行回答：
- 自身总结：您自己在本周课程的学习，收获，技能掌握等方面进行总结，包括自身在哪些方面存在哪些不足，欠缺，困惑。作为将来回顾学习路径时的依据。
- 课程反馈：也可以就知识点，进度，难易度，教学方式，考试方式等等进行意见反馈，督促我们进行更有效的改进，为大家提供更优质的服务。